File: gp_app/lib/models/chat_message.dart
// import 'package:flutter/material.dart';

class ChatMessage {
  final int senderId;
  final int receiverId;
  final int? burnId;
  final String message;
  final String? image; // Updated to store base64 encoded image
  final int? imgFlag;
  final String? voiceNote;
  final bool? receiver;
  final DateTime timestamp;
  // Variables Related to the Location
  final double? latitude; // Add latitude
  final double? longitude; // Add longitude
  final String? hospitalNameEn;
  final String? hospitalNameAr;
  final List<Map<String, String>>? hospitalDetails;

  ChatMessage({
    required this.senderId,
    required this.receiverId,
    this.burnId,
    this.imgFlag,
    required this.message,
    required this.receiver,
    this.image,
    this.voiceNote,
    required this.timestamp,
    // Variables Related to the Location
    this.latitude, // Initialize latitude
    this.longitude, // Initialize longitude
    this.hospitalNameEn,
    this.hospitalNameAr,
    this.hospitalDetails,
  });

  // Factory constructor to create a ChatMessage from JSON
  factory ChatMessage.fromJson(Map<String, dynamic> json) {
    return ChatMessage(
      senderId: json['sender_id'] as int,
      receiverId: json['receiver_id'] as int,
      burnId: json['burn_id'] as int?,
      message: json['message'] as String? ?? '',
      image: json['image'] as String?, // Store base64 encoded image
      imgFlag: json['img_flag'] as int?,
      voiceNote: json['voice_note_path'] as String?,
      receiver: json['receiver'] as bool?,
      timestamp: DateTime.parse(json['timestamp'] as String),
      latitude: json['latitude'] as double?,
      longitude: json['longitude'] as double?,
      hospitalNameEn: json['hospital_name_en'] as String?,
      hospitalNameAr: json['hospital_name_ar'] as String?,
      hospitalDetails: (json['hospital_details'] as List<dynamic>?)
          ?.cast<Map<String, String>>(),
    );
  }

  // Convert instance back to JSON, useful for sending data back to a server or saving locally
  Map<String, dynamic> toJson() {
    return {
      'sender_id': senderId,
      'receiver_id': receiverId,
      'burn_id': burnId,
      'message': message,
      'image': image, // Store base64 encoded image
      'voice_note': voiceNote,
      'receiver': receiver,
      'timestamp': timestamp.toIso8601String(),
      'latitude': latitude,
      'longitude': longitude,
      'img_flag': imgFlag,
      'hospital_name_en': hospitalNameEn,
      'hospital_name_ar': hospitalNameAr,
      'hospital_details': hospitalDetails,
    };
  }
}

File: gp_app/Backend/chat_message.py
from flask_sqlalchemy import SQLAlchemy
from .base import db


class ChatMessage(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    sender_id = db.Column(db.Integer, nullable=False)
    receiver_id = db.Column(db.Integer, nullable=False)
    message = db.Column(db.String(2000), nullable=True)
    image =  db.Column(db.LargeBinary(), nullable=True)
    img_flag = db.Column(db.Integer, nullable=True)
    receiver = db.Column(db.Boolean, nullable=False, default=False)
    timestamp = db.Column(db.DateTime, server_default=db.func.now())
    burn_id = db.Column(db.Integer, nullable=False)
    # user_id = db.Column(db.Integer, nullable=False)
    show_btn = db.Column(db.Boolean, nullable=True, default=False)
    voice_note_path = db.Column(db.String(255), nullable=True) 

    def to_dict(self):
        return {
            'id': self.id,
            'sender_id': self.sender_id,
            'receiver_id': self.receiver_id,
            'message': self.message,
            'image': self.image,
            'img_flag': self.img_flag,
            'receiver': self.receiver,
            'show_btn': self.show_btn,
            'timestamp': self.timestamp.isoformat(),
            'burn_id': self.burn_id,
            #'user_id': self.user_id,
            'voice_note_path': self.voice_note_path,
        }
